modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/exp/IgniteTableFunction.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.ignite.internal.processors.query.calcite.exec.exp;
+
+import java.lang.reflect.Method;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.apache.calcite.adapter.enumerable.NullPolicy;
+import org.apache.calcite.adapter.java.JavaTypeFactory;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeFactory;
+import org.apache.calcite.schema.TableFunction;
+import org.apache.ignite.cache.query.annotations.QuerySqlFunction;
+import org.apache.ignite.internal.processors.query.IgniteSQLException;
+import org.apache.ignite.internal.util.typedef.F;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Holder of user-defined table function.
+ *
+ * @see QuerySqlFunction#tableColumnTypes()
+ */
+public class IgniteTableFunction extends IgniteReflectiveFunctionBase implements TableFunction {
+    /** Column types of the returned table representation. */
+    private final Class<?>[] colTypes;
+
+    /** Column names of the returned table representation. */
+    private final List<String> colNames;
+
+    /**
+     * Creates user-defined table function holder.
+     *
+     * @param method The implementatng method.
+     * @param colTypes Column types of the returned table representation.
+     * @param colNames Column names of the returned table representation. Of empty, the defaults are used ('COL_1', 'COL_2', etc.).
+     * @param implementor Call implementor.
+     */
+    private IgniteTableFunction(Method method, Class<?>[] colTypes, @Nullable String[] colNames, CallImplementor implementor) {
+        super(method, implementor);
+
+        validate(method, colTypes, colNames);
+
+        this.colTypes = colTypes;
+
+        this.colNames = F.isEmpty(colNames) ? new ArrayList<>(colTypes.length) : Arrays.asList(colNames);
+
+        if (F.isEmpty(colNames)) {
+            for (int i = 0; i < colTypes.length; ++i)
+                this.colNames.add("COL_" + (i + 1));
+        }
```

Comment:
1. Not sure we should allow column names auto-generation
2. If use auto-generation, we should do it in Calcite style (see SqlUtil#GENERATED_EXPR_ALIAS_PREFIX)